ads/google/csa.js

```diff
@@ -0,0 +1,264 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {validateData, loadScript} from '../../3p/3p';
+import {tryParseJson} from '../../src/json.js';
+
+// Keep track of current height of AMP iframe
+let currentAmpHeight = null;
+
+/**
+ * Request Custom Search Ads (Adsense for Search or AdSense for Shopping).
+ * @param {!Window} global The window object of the iframe
+ * @param {!Object} data
+ */
+export function csa(global, data) {
+  // Get parent width in case we want to override
+  const width = global.document.body.clientWidth;
+
+  validateData(data, [], [
+    'afshPageOptions',
+    'afshAdblockOptions',
+    'afsPageOptions',
+    'afsAdblockOptions',
+    'ampSlotIndex',
+  ]);
+
+  // Add the ad container to the document
+  const containerDiv = global.document.createElement('div');
+  const containerId = 'csacontainer';
+  containerDiv.id = containerId;
+  global.document.body.appendChild(containerDiv);
+
+  // Parse all the options
+  const pageOptions = {source: 'amp', referer: global.context.referrer};
+  const adblockOptions = {container: containerId};
+  const afshPageOptions = Object.assign(
+      Object(tryParseJson(data['afshPageOptions'])), pageOptions);
```

Comment:
did you try swap the argument position?  That will help you get rid of `Object()`.

```
Object.assign(pageOptions, tryParseJson(data['afshPageOptions']));
```